Update covidhub import to handle internal CZBID

ttung opened the pull request:
### Description
This adds support for Internal CZBIDs.

The old workflow was:

1. look at all consensus genomes.
2. look at all CZBIDs that are not represented by the consensus genomes.
3. write out entries for the consensus genomes.
4. write out entries for the failed genomes.

This refactors the workflow to:

1. Pull all the CZBIDs associtaed with the project.
2. Load all the consensus genomes.
3. Loop through all the CZBIDs and:
   a. Find the consensus genome
   b. Sort out the metadata, depending on the CZBID type.

Depends on #297 

#### Issue
[ch133318](https://app.clubhouse.io/genepi/story/133318/port-over-internal-samples-from-covidhub-db-into-appropriate-projects-about-500)

### Test plan
import marin's internal czbids (RR089i)

phoenixAja commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +152,196 @@ def import_project(
+                internal_metadata = internal_czb_ids_metadata[czbid.czb_id]
+                external_accession = internal_metadata[0]
+                collection_date = internal_metadata[1]
+                date_received = internal_metadata[1]
```
should this be this? 
```suggestion
                date_received = internal_metadata[2]
```

ttung replied:
no, this is intentionally the same.  we don't really know the collection date, so we're setting it to `date_received`.

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -103,8 +120,23 @@ def import_project(
+        group_czbids: Collection[CZBID] = (
+            covidhub_session.query(CZBID)
+            .join(Project)
+            .filter(Project.rr_project_id == rr_project_id)
```
So the rr_project id is different for internal/external samples though, ex. `RR0092i` vs `RR0092e` am I reading this wrong or does it assume they are the same?

ttung replied:
We'd run the import twice -- once for `RR0092e` and once for `RR0092i`

shanaxel42 replied:
ahhhhh then all good!

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -103,8 +120,23 @@ def import_project(
+        )
+
+        czbid_lookup_by_str: Mapping[str, CZBID] = {
+            czbid.czb_id: czbid for czbid in group_czbids
```
Why do we do this str comparison?

ttung replied:
There is no string comparison...?
This is to look up a czbid based on the string.

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +152,196 @@ def import_project(
-                    }
-                )
+            .join(CZBIDRnaPlate)
+            .join(
```
clleeeeeean

ttung replied:
this should just be one join.

ttung replied:
fixed!

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +152,196 @@ def import_project(
+        czbids_to_process: MutableSequence[CZBID] = list()
+        for external_accession, czbids in external_accession_to_czbids.items():
+            if len(czbids) > 1:
+                # gather up the information for each czbid representing the external
```
So I'm not sure we want this logic. If a county accidentally sent two samples with the same external_accession they may actually be two separate samples. Ideally there wouldn't be any of these, but are there?

ttung replied:
Yes, there are.

These accessions are assigned when czb got the tubes (`AccessionSamples`), and not assigned by the counties afaik.

ttung replied:
Example for Marin:

```
External accession H12313 represented by multiple czbids: {'RR089i_00186', 'RR089i_00181'}.  Choosing RR089i_00181 as the czbid forthis sample.
External accession H11851 represented by multiple czbids: {'RR089i_00182', 'RR089i_00187'}.  Choosing RR089i_00182 as the czbid forthis sample.
```

shanaxel42 replied:
Oh interesting these are internal samples that were picked twice, ok yea then this makes sense

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +152,196 @@ def import_project(
+            if isinstance(czbid, DphCZBID):
+                if czbid.external_accession in (
+                    "OCPHL2207",
+                    "OCPHL2208",
```
lolz

shanaxel42 reviewed (approved):
Looks good!

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +152,196 @@ def import_project(
+            czbid: (external_accession, collection_date)
+            for czbid, external_accession, collection_date in covidhub_session.query(
+                InternalCZBID.czb_id.label("czb_id"),
+                AccessionSample.accession.label("external_accession"),
```
OK! so yes we want this to be

`AccessionSample.location_submitter_id.label("external_accession"),`

Commit:
Update covidhub import to handle internal CZBID

### Description
This adds support for Internal CZBIDs.

The old workflow was:

1. look at all consensus genomes.
2. look at all CZBIDs that are not represented by the consensus genomes.
3. write out entries for the consensus genomes.
4. write out entries for the failed genomes.

This refactors the workflow to:

1. Pull all the CZBIDs associtaed with the project.
2. Load all the consensus genomes.
3. Loop through all the CZBIDs and:
   a. Find the consensus genome
   b. Sort out the metadata, depending on the CZBID type.

There is an additional wrinkle in how we handle internal czbids, as some samples have multiple CZBIDs.  This can happen if they are sequenced multiple times.  In this case, we pick the CZBID that has the most recovered sites, and if there is a tie there, then we pick the most recently sequenced CZBID.

#### Issue
[ch133318](https://app.clubhouse.io/genepi/story/133318/port-over-internal-samples-from-covidhub-db-into-appropriate-projects-about-500)

### Test plan
import marin's internal czbids (RR089i)

shanaxel42 commented on src/backend/aspen/covidhub_import/implementation.py:
```diff
@@ -120,68 +166,195 @@ def import_project(
+            )
+            for czbid, external_accession, collection_date in covidhub_session.query(
+                InternalCZBID.czb_id.label("czb_id"),
+                AccessionSample.location_submitter_id.label("external_accession"),
```
nice!

shanaxel42 reviewed (approved):
Looks good!
